sopel/irc/backends.py

```diff
@@ -5,346 +5,330 @@
-                                       keyfile=self.keyfile,
-                                       do_handshake_on_connect=True,
-                                       suppress_ragged_eofs=True,
-                                       cert_reqs=ssl.CERT_REQUIRED,
-                                       ca_certs=self.ca_certs)
-            # connect to host specified in config first
-            try:
-                ssl.match_hostname(self.ssl.getpeercert(), self.host)
-            except ssl.CertificateError:
-                # the host in config and certificate don't match
-                LOGGER.error("hostname mismatch between configuration and certificate")
-                # check (via exception) if a CNAME matches as a fallback
-                has_matched = False
-                for hostname in get_cnames(self.host):
-                    try:
-                        ssl.match_hostname(self.ssl.getpeercert(), hostname)
-                        LOGGER.warning(
-                            "using {0} instead of {1} for TLS connection"
-                            .format(hostname, self.host))
-                        has_matched = True
-                        break
-                    except ssl.CertificateError:
-                        pass
-
-                if not has_matched:
-                    # everything is broken
-                    LOGGER.error("Invalid certificate, no hostname matches.")
-                    # TODO: refactor access to bot's settings
-                    if hasattr(self.bot.settings.core, 'pid_file_path'):
-                        # TODO: refactor to quit properly (no "os._exit")
-                        os.unlink(self.bot.settings.core.pid_file_path)
-                        os._exit(1)
-        self.set_socket(self.ssl)
-        LOGGER.info('Connection accepted by the server...')
-        LOGGER.debug('Starting job scheduler for connection timeout...')
-        self.timeout_scheduler.start()
+            self.bot.on_message(data)
+
+        # cancel timeout tasks
+        self._cancel_timeout_tasks()
+
+    # run & connection
+
+    def get_connection_kwargs(self) -> Dict:
+        """Return the keyword arguments required to initiate connection."""
+        ssl_context: Optional[ssl.SSLContext] = None
+
+        if self._use_ssl:
+            ssl_context = ssl.create_default_context(ssl.Purpose.SERVER_AUTH)
+            if self._certfile is not None:
+                # load_cert_chain requires a certfile (cannot be None)
+                ssl_context.load_cert_chain(
+                    certfile=self._certfile,
+                    keyfile=self._keyfile,
+                )
+
+            if not self._verify_ssl:
+                # deactivate SSL verification for hostname & certificate
+                ssl_context.check_hostname = False
+                ssl_context.verify_mode = ssl.CERT_NONE
+            else:
+                # verify location if verify_mode is not CERT_NONE
```

Comment:
As said on IRC, it all comes to the default value for both `verify_mode` and `check_hostname` based on `purpose` being `SERVER_AUTH`. Moreover, if you have `check_hostname=True`, then you cannot have something different than `verify_mode` set to `CERT_REQUIRED`.

Also, there is only 3 possible values:

* `CERT_NONE`: must be manually set
* `CERT_REQUIRED`: set by default on `SERVER_AUTH`
* `CERT_OPTIONAL`: same as `CERT_REQUIRED` for client context (which is the case for the purpose of `SERVER_AUTH`); otherwise allow the client to refuse the certificate (in server context)

So it's a comment that implies that you know a bit about how `ssl.create_default_context(ssl.Purpose.SERVER_AUTH)` works.

Reply:
I could make a comment here about following the Zen of Python and setting explicit values even if they're identical to the defaults… But really, it's just the wording of the comment itself that bugs me, because I read about those defaults recently while looking at the other TLS PR and the comment still didn't make sense.

Maybe reordering the if/else conditions would make more sense: If verify_ssl, set that up; else disable verification in the SSL context. Would you be OK with that?

No suggestion, sorry. I'm on my phone, waiting for AAA to rescue our car that won't start. 🙃

Reply:
I think it's clear enough what's happening without the comment, or even better with the conditional flipped.